File: 207-course-schedule/course-schedule.cpp
class Solution {
public:

    bool dfs(int node,unordered_map<int, vector<int>> & adj,vector<int> & visited,set<int> &parent )
    {
        
        
        visited[node] = 1;
        parent.insert(node);

        for(auto neig : adj[node])
        {  
            if(parent.count(neig)) return false;
            
            if(!visited[neig])
            {
                if(!dfs(neig, adj, visited, parent))
                {
                    return false;
                }
            }
           
        }
         parent.erase(node);
        
        return true;
    }


    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>>adj;
        for(auto & pre : prerequisites)
        {
            adj[pre[0]].push_back(pre[1]);
        }

        vector<int> visited(numCourses,0);
        set<int> parent;
        bool result = false;
        for(int i = 0;i<numCourses;i++)
        {
            if(!visited[i])
            {
                if(!dfs(i , adj, visited, parent)){
                    return false;
                }
            }
        }

        return true;
        
    }
};

File: 207-course-schedule/course-schedule.java
class Solution {

    // boolean dfs(int curr, int[] visited, List<List<Integer>> adj)
    // {

    //     if(visited[curr] == 1) return true;
    //     if(visited[curr] == 2) return false;

    //     visited[curr] = 1;

    //     for(int neigh : adj.get(curr))
    //     {
            

           
    //             if(dfs(neigh, visited, adj)){
    //                 return true;
    //             }
            
    //     }
    //     visited[curr] =2;
    //     return false;

    // }



    public boolean canFinish(int numCourses, int[][] prerequisites) {

        int[] indegree = new int[numCourses];
        List<List<Integer>> adj = new ArrayList<>();
        for(int i = 0; i < numCourses ; i++)
        {
            adj.add(new ArrayList<>());
        }
        for(int[] edge : prerequisites)
        {
            adj.get(edge[1]).add(edge[0]);
            indegree[edge[0]]++;
        }
        Queue<Integer> q = new LinkedList<>();

        for(int i = 0; i< numCourses; i++)
        {
           
            if(indegree[i] == 0)
            {
                q.offer(i);
            }
        }
        List<Integer> topo = new ArrayList<>();

        while(!q.isEmpty())
        {
            int curr = q.poll();
            topo.add(curr);

           
            for(int neig : adj.get(curr))
            {
                indegree[neig]--;
                if(indegree[neig] == 0) {
                    q.offer(neig);
                    
                }
            }
        }

        return topo.size() == numCourses;
        




        // List<List<Integer>> adj =new ArrayList<>();
        // for(int i = 0;i <numCourses; i++)
        // {
        //     adj.add(new ArrayList<>());
        // }

        // for(int [] edge : prerequisites)
        // {
        //     adj.get(edge[1]).add(edge[0]);
        // }

        // int[] visited = new int[numCourses];
        // for(int i = 0 ;i < numCourses; i++)
        // {
        //     if(dfs(i,visited, adj)){
        //     return false;
        // }
        // }
        

        // return true;
    }
}